### ISSUE TITLE
Java 9 module-info.java ruminations

### ISSUE THREAD

**Issue by raelik** [jruby/jruby#4835]
After Java 9 compatibility is made tentatively available in 9.1.14, discussions need to be had on how best to set up jRuby's module-info.java file(s) for the Java Platform Module System now available in Java 9.

**Comment by headius** [comment]
We will also need to modularize the libraries we package, including but not limited to:

* All the JNR libraries
* bytelist/jcoding/joni
* utility and JVM shim libraries (invokebinder already done)

We also package several third-party libs and will want to see if they've started modularizing.

**Comment by headius** [comment]
Tentatively marked for 9.2.

One big wrinkle here is that if we start including module-info in a project, we're going to need Java 9 tooling to build it. That's not a huge leap for some of the libraries, since we control them and are generally the only ones maintaining them, but JRuby itself is built by and for many different systems, some of which may not have Java 9 for a while.

**Comment by raelik** [comment]
@headius Maybe just a separate build target for a modularized Java 9 deployment that includes the module-info, akin to how there is currently binary, Windows executable, source and complete jar packages of the jRuby release.

**Comment by headius** [comment]
I'm trying to wrap up my day, but perhaps @raelik or someone else observing this issue can research whether there are any ways to build module-info.java without a Java 9 install present. I know we can make this produce a jar that works on Java 7 or 8, as I did in https://github.com/headius/invokedynamic.

**Comment by raelik** [comment]
@headius as far as I can see, the syntax of the module-info.java file is only understood by Java 9's `javac`.

**Comment by headius** [comment]
@raelik Yeah, I'm just wondering if there's a way to pull down OpenJDK 9's javac or an equivalent javac that understand module-info.java without running on 9.

**Comment by headius** [comment]
We will put more focus into module support over the summer in preparation for the Java 11 LTS release.

**Comment by headius** [comment]
Some of our deps have been modularized: jcodings, joni. The bytelist library has been rolled back into JRuby proper due to duplicated packages.

The jnr projects still need to be modularized.

I added an automatic module name in 681ab2210930854e7fbd8bb5d8f728f205363554.

More to come, ideally with full proper modularization in 9.2.2.

**Comment by headius** [comment]
Much of this has been done. Fully modularizing JRuby and all dependencies will be a 9.3 task.

**Comment by headius** [comment]
As of this writing, the following libraries either have no module info or we do not know their module status:

* org.ow2.asm:asm, -commons, -util, unknown status but recent releases most likely have module info
* org.jruby:dirgra owned by @enebo 
* com.headius:options owned by @headius 
* bsf:bsf unknown status
* com.jcraft:jzlib unknown status
* com.martiansoftware:nailgun-server
* org.apache.ant:ant unknown status
* org.osgi:org.osgi.core unknown status but likely modularized
* joda-time:joda-time unknown status but likely modularized
* org.slf4j:slf4j-api and -simple, unknown status
* me.qmx.jitescript:jitescript, formerly owned by @qmx, unknown status but no releases in a long time and unlikely to be modularized
* com.headius:backport9 owned by @headius

The remaining libraries known to provide module info currently do so through Automatic-Module-Name, which means they are not fully modularized (defaulting to everything open and not specifying system deps).

Other decisions that we need to make on this:

* Shaded JRuby is now primarily used only in the distribution, since we introduced the no-shading jruby-base module. However modularity will want us to drop all shading since multiple modules cannot be combined into a single jar in this way. It may be in our best interest to ship the JRuby distribution as the jruby-base jar plus dependency jars loose on the filesystem and configured into the JDK module-path.
* JRuby itself does provide an automatic module name but currently no module-info that would specify dependencies and exports. This may not be possible until the other dependencies are all modularized.

There's a good chance this will slip from 9.3 but I will at least try to track down modularized versions of these deps.

**Comment by headius** [comment]
com.headius.invokebinder does have a full module-info.

**Comment by headius** [comment]
bsf:bsf has no newer release, but we do not ship it in any form. It is only used for building our BSF support.

**Comment by headius** [comment]
#6596 removes BSF entirely to dodge that whole issue.

**Comment by headius** [comment]
jzlib appears to no longer be maintained. I am trying to ascertain its status, but we may be forced to take over the library (perhaps with the help of others) in order to modularize it.

https://github.com/ymnk/jzlib/issues/25

I believe for libraries that do not have module support, we could just continue shading them... but ideally we would have proper modules up and down the JRuby dependency stack.

**Comment by headius** [comment]
Missed dependency: com.martiansoftware:nailgun-server

I would propose that we finally remove the rest of Nailgun support in 9.3.

**Comment by headius** [comment]
Nailgun support was partially removed in https://github.com/jruby/jruby/issues/5981 and the rest is removed in #6597.

**Comment by headius** [comment]
joda-time has module support and we are on the latest.

**Comment by headius** [comment]
org.apache.ant:ant and org.osgi:org.osgi.core are not shipped with JRuby and may not need modularity. Unsure how sources in JRuby that optionally depend on osgi will work in a module environment.

**Comment by slonopotamus** [comment]
Why JRuby cannot just use `java.util.zip.Deflater`/`java.util.zip/Inflater` instead of JZLib?

**Comment by headius** [comment]
@slonopotamus I don't recall the exact features, but there are some low-level aspects of zlib exposed by the Ruby library that we could not get from the built-in deflater and inflater. It is possible those features have been exposed in the decade since we switched to jzlib, but I have not done that audit.

If we can eliminate jzlib that would be a positive... not because there's anything wrong with it as a library but because it would be one less dependency.

**Comment by headius** [comment]
@slonopotamus I was curious so I dug back to this commit, which was the first move to JZlib: 908cd0818635a09883db397a05469989df138908

The commits this squashed explain some of the behaviors we could not get out of the default JDK library, such as:

* Various specific error conditions hidden behind java.util.zip's generic exceptions
* Methods that could not be supported: GzipReader#unused, GzipWriter#sync, GzipWriter#sync=, GzipWriter#flush

It would be great if we could do everything with java.util.zip (if possible) but it is not a priority for us right now. We would happily support anyone that wanted to attempt it!

**Comment by headius** [comment]
Linking to #6598 which has been punted to 9.4. The issues are listed there, but most of the to-do's in this issue are resolved. Nearly all libraries have at least an automatic module name, and we will try to get full descriptors into them for 9.4. Unfortunately [JDK-8263432](https://bugs.openjdk.java.net/browse/JDK-8263432) prevents us from doing this modularization move in 9.3.

**Comment by headius** [comment]
This one needs to keep getting kicked down the road. We have made great improvements to module support in 9.3 and 9.4, but we still cannot build cleanly with modules active and we have several libraries that attempt to cross module boundaries illegally.

Punting to 9.4.1, because we really need to fix this before 9.5.

**Comment by headius** [comment]
Remaining work needs to happen in 9.5 but we are functioning well under all JPMS versions of JDK currently.
